### PR TITLE
refactor(typescript): remove unused locals and lint more modules

### PR DESCRIPTION
**Pull request by alfaproject**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[x] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
No CI checks exist for unused locals or variables.

**What is the new behavior?**
TypeScript compiler will check for unused locals in some projects.
TSLint is now configured to flag unused variables, and is extended to more projects.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

**Other information**:
Ideally, the Typescript compiler should be used to check for unused locals in all projects, but some of the current projects include generated code, and this PR is already big enough.

P.S.: I might have found a bug with the XTB loading.

### CODE REVIEW

**Comment by alfaproject** [comment]
I'm leaving this PR as WIP for now, because I want to double check tests, and maybe get some feedback.

File: gulpfile.js [hunk 1]
```diff
@@ -148,14 +148,15 @@ gulp.task('lint', ['check-tests', 'format:enforce', 'tools:build'], () => {
       .src([
         // todo(vicb): add .js files when supported
         // see https://github.com/palantir/tslint/pull/1515
-        'modules/@angular/**/*.ts',
```
Comment (vicb) [inline gulpfile.js]: why ?

File: modules/@angular/compiler-cli/src/static_reflector.ts [hunk 2]
```diff
@@ -255,7 +255,6 @@ export class StaticReflector implements ReflectorReader {
       }
 
       function resolveReferenceValue(staticSymbol: StaticSymbol): any {
-        let result: any = staticSymbol;
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/static_reflector.ts]: I think we are both changing the same code across different PRs...
One of us will need to rebase !

File: modules/@angular/compiler/src/compiler.ts [hunk 3]
```diff
@@ -124,11 +124,12 @@ function _initReflector() {
  *
  * @experimental
  */
-export const platformCoreDynamic = createPlatformFactory(platformCore, 'coreDynamic', [
-  {provide: COMPILER_OPTIONS, useValue: {}, multi: true},
-  {provide: CompilerFactory, useClass: RuntimeCompilerFactory},
-  {provide: PLATFORM_INITIALIZER, useValue: _initReflector, multi: true},
-]);
+export const platformCoreDynamic: (extraProviders?: Provider[]) => PlatformRef =
```
Comment (vicb) [inline modules/@angular/compiler/src/compiler.ts]: why ?

File: modules/@angular/compiler/src/i18n/serializers/xtb.ts [hunk 4]
```diff
@@ -179,8 +179,7 @@ class _Visitor implements ml.Visitor {
 
   visitExpansion(expansion: ml.Expansion, context: any): any {
     const strCases = expansion.cases.map(c => c.visit(this, null));
-
-    return `{${expansion.switchValue}, ${expansion.type}, strCases.join(' ')}`;
+    return `{${expansion.switchValue}, ${expansion.type}, ${strCases.join(' ')}}`;
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/serializers/xtb.ts]: extract this in a separate "fix" commit, add a test

File: modules/@angular/compiler/test/i18n/integration_spec.ts [hunk 5]
```diff
@@ -177,27 +177,38 @@ const XTB = `
 // unused, for reference only
 // can be generated from xmb_spec as follow:
 // `iit('extract xmb', () => { console.log(toXmb(HTML)); });`
+/*
```
Comment (vicb) [inline modules/@angular/compiler/test/i18n/integration_spec.ts]: what is this

File: modules/@angular/platform-browser-dynamic/src/platform-browser-dynamic.ts [hunk 6]
```diff
@@ -23,5 +23,6 @@ export const RESOURCE_CACHE_PROVIDER: Provider[] =
 /**
  * @stable
  */
-export const platformBrowserDynamic = createPlatformFactory(
-    platformCoreDynamic, 'browserDynamic', INTERNAL_BROWSER_DYNAMIC_PLATFORM_PROVIDERS);
+export const platformBrowserDynamic: (extraProviders?: Provider[]) => PlatformRef =
```
Comment (vicb) [inline modules/@angular/platform-browser-dynamic/src/platform-browser-dynamic.ts]: why ?

File: modules/@angular/platform-browser/src/security/url_sanitizer.ts [hunk 7]
```diff
@@ -39,9 +39,6 @@ import {getDOM} from '../dom/dom_adapter';
  */
 const SAFE_URL_PATTERN = /^(?:(?:https?|mailto|ftp|tel|file):|[^&:/?#]*(?:[/?#]|$))/gi;
 
-/* A pattern that matches safe srcset values */
```
Comment (vicb) [inline modules/@angular/platform-browser/src/security/url_sanitizer.ts]: please isolate this in a separate commit (as it touches security)

File: modules/@angular/platform-browser/test/testing_public_spec.ts [hunk 8]
```diff
@@ -402,7 +402,6 @@ export function main() {
 
     describe('errors', () => {
       var originalJasmineIt: any;
-      var originalJasmineBeforeEach: any;
```
Comment (vicb) [inline modules/@angular/platform-browser/test/testing_public_spec.ts]: not sure about this one, please isolate in a sperate PR

File: modules/benchmarks/e2e_test/old/naive_infinite_scroll_spec.ts [hunk 9]
```diff
@@ -1,5 +1,11 @@
+/**
```
Comment (vicb) [inline modules/benchmarks/e2e_test/old/naive_infinite_scroll_spec.ts]: this should be tested, is this related to your gulpfile updates ?

isolate in a separate commit/PR

**Comment by vicb** [comment]
Thanks for the cleanup / updates. That's a very valuable PR.

However I gave up reviewing half way because the changes are too many.

Please take the comments into account and split this PR into multiple commmits / PRs where it makes sense (by component / type of change, ...)

Thanks

File: gulpfile.js [hunk 10]
```diff
@@ -148,14 +148,15 @@ gulp.task('lint', ['check-tests', 'format:enforce', 'tools:build'], () => {
...
+        'modules/**/*.ts',
+        'tools/**/*.ts',
         './*.ts',
+        '!**/*.d.ts',
+        '!**/*.ngfactory.ts',
       ])
       .pipe(tslint({
         tslint: require('tslint').default,
         configuration: tslintConfig,
-        rulesDirectory: 'dist/tools/tslint',
```
Comment (alfaproject) [inline gulpfile.js]: I moved it to tslint confit, so it can also be used by IDE tslint plugins.

File: gulpfile.js (hunk 1 above)
Comment (alfaproject) [inline gulpfile.js]: All modules are linted now. Can't see why not, but I can revert if it's preferable.

File: modules/@angular/benchpress/src/sample_description.ts [hunk 11]
```diff
@@ -30,8 +30,9 @@ export class SampleDescription {
                 ],
                 metric.describe()),
     deps: [
-      Metric, Options.SAMPLE_ID, Options.FORCE_GC, Options.USER_AGENT, Validator,
-      Options.DEFAULT_DESCRIPTION, Options.SAMPLE_DESCRIPTION
+      Metric, <OpaqueToken>Options.SAMPLE_ID, <OpaqueToken>Options.FORCE_GC,
```
Comment (alfaproject) [inline modules/@angular/benchpress/src/sample_description.ts]: Work around tslint not knowing that those imports are being used.

File: modules/@angular/compiler-cli/src/static_reflector.ts (hunk 2 above)
Comment (alfaproject) [inline modules/@angular/compiler-cli/src/static_reflector.ts]: No worries, I can rebase. Had to make the change to pass linting tests anyway.

File: modules/@angular/compiler/test/i18n/integration_spec.ts (hunk 5 above)
Comment (alfaproject) [inline modules/@angular/compiler/test/i18n/integration_spec.ts]: Unused variable. Can either remove or comment out. Left it commented for now.

File: modules/@angular/forms/src/directives/select_multiple_control_value_accessor.ts [hunk 12]
```diff
@@ -14,8 +14,8 @@ import {isBlank, isPresent, isPrimitive, looseIdentical} from '../facade/lang';
 import {ControlValueAccessor, NG_VALUE_ACCESSOR} from './control_value_accessor';
 
 export const SELECT_MULTIPLE_VALUE_ACCESSOR = {
-  provide: NG_VALUE_ACCESSOR,
-  useExisting: forwardRef(() => SelectMultipleControlValueAccessor),
+  provide: <OpaqueToken>NG_VALUE_ACCESSOR,
```
Comment (alfaproject) [inline modules/@angular/forms/src/directives/select_multiple_control_value_accessor.ts]: Tslint blindness

File: modules/@angular/platform-browser-dynamic/src/platform-browser-dynamic.ts (hunk 6 above)
Comment (alfaproject) [inline modules/@angular/platform-browser-dynamic/src/platform-browser-dynamic.ts]: Tslint blindness

File: modules/@angular/platform-browser/src/security/url_sanitizer.ts (hunk 7 above)
Comment (alfaproject) [inline modules/@angular/platform-browser/src/security/url_sanitizer.ts]: It's not being used tho. Guess I can comment it out as an alternative?

File: modules/@angular/platform-browser/test/testing_public_spec.ts (hunk 8 above)
Comment (alfaproject) [inline modules/@angular/platform-browser/test/testing_public_spec.ts]: This is not being used, had to remove to pass linting. Guess I can comment tho

File: modules/benchmarks/e2e_test/old/naive_infinite_scroll_spec.ts (hunk 9 above)
Comment (alfaproject) [inline modules/benchmarks/e2e_test/old/naive_infinite_scroll_spec.ts]: Just added the header

**Comment by alfaproject** [comment]
Fair enough. Changing the gulpfile to extend linting to more source files will inevitably create a big PR or else CI will fail, but I'll make the changes in separate commits. Will need a few days tho

File: modules/@angular/compiler/src/compiler.ts (hunk 3 above)
Comment (vicb) [inline modules/@angular/compiler/src/compiler.ts]: Please make this change part of a separate commit then. Thanks.

File: modules/@angular/compiler/test/i18n/integration_spec.ts (hunk 5 above)
Comment (vicb) [inline modules/@angular/compiler/test/i18n/integration_spec.ts]: pb with ts lint ? if yes, deactivate the check for this code block

File: modules/@angular/platform-browser/test/testing_public_spec.ts (hunk 8 above)
Comment (vicb) [inline modules/@angular/platform-browser/test/testing_public_spec.ts]: > I can comment

commenting is useless. Should be deleted if not needed - but please not in this file for now.

**Comment by alfaproject** [comment]
I'll be creating separate PRs with separate commits so that reviewing easier and avoid too many conflicts with other open PRs.
